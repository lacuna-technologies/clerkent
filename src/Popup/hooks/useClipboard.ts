import { useState, useCallback, useEffect } from 'preact/hooks'
import { browser } from 'webextension-polyfill-ts'

const useClipboard = () => {
  const [permissionGranted, setPermissionGranted] = useState(false)

  const promptGrant = useCallback(async () => {
    const permissionsRequest = await browser.permissions.request({
      permissions: [`clipboardRead`],
    })
    setPermissionGranted(permissionsRequest)
  }, [])

  useEffect(() => {
    (async () => {
      const granted = await browser.permissions.contains({ permissions: [`clipboardRead`] })
      setPermissionGranted(granted)
    })()
  })

  return {
    permissionGranted,
    promptGrant,
  }
}

export default useClipboard