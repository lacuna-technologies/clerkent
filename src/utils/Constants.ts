import type Law from '../types/Law'

const JURISDICTIONS = {
  AU: {
    emoji: `ðŸ‡¦ðŸ‡º`,
    id: `AU` as Law.JursidictionCode,
    name: `Australia`,
  },
  CA: {
    emoji: `ðŸ‡¨ðŸ‡¦`,
    id: `CA` as Law.JursidictionCode,
    name: `Canada`,
  },
  EU: {
    emoji: `ðŸ‡ªðŸ‡º`,
    id: `EU` as Law.JursidictionCode,
    name: `European Union`,
  },
  HK: {
    emoji: `ðŸ‡­ðŸ‡°`,
    id: `HK` as Law.JursidictionCode,
    name: `Hong Kong`,
  },
  MY: {
    emoji: `ðŸ‡²ðŸ‡¾`,
    id: `MY` as Law.JursidictionCode,
    name: `Malaysia`,
  },
  NZ: {
    emoji: `ðŸ‡³ðŸ‡¿`,
    id: `NZ` as Law.JursidictionCode,
    name: `New Zealand`,
  },
  SG: {
    emoji: `ðŸ‡¸ðŸ‡¬`,
    id: `SG` as Law.JursidictionCode,
    name: `Singapore`,
  },
  UK: {
    emoji: `ðŸ‡¬ðŸ‡§`,
    id: `UK` as Law.JursidictionCode,
    name: `United Kingdom`,
  },
}


const SG_DATABASES = {
  lawnet: {
    icon: ``,
    name: `Lawnet`,
    url: `https://www.lawnet.sg/lawnet/web/lawnet/home`,
  },
  lexis: {
    icon: ``,
    name: `Lexread`,
    url: `https://www.lexread.lexisnexis.com/`,
  },
  sc: {
    icon: ``,
    name: `Supreme Court`,
    url: `https://www.supremecourt.gov.sg/news/supreme-court-judgments`,
  },
  slw: {
    icon: ``,
    name: `Singapore Law Watch`,
    url: `https://www.singaporelawwatch.sg/Judgments`,
  },
  sso: {
    icon: ``,
    name: `Singapore Statutes Online`,
    url: `https://sso.agc.gov.sg/`,
  },
  westlaw: {
    icon: ``,
    name: `Westlaw Asia`,
    url: `https://www.westlawasia.com/singapore/`,
  },
}

const UK_DATABASES = {
  bailii: {
    icon: ``,
    name: `BAILII`,
    url: `https://bailii.org/`,
  },
  legislation: {
    icon: ``,
    name: `legislation.gov.uk`,
    url: `https://www.legislation.gov.uk/`,
  },
  lexis: {
    icon: ``,
    name: `LexisNexis`,
    url: `http://www.lexisnexis.com/uk/legal`,
  },
  westlaw: {
    icon: ``,
    name: `Westlaw`,
    url: `https://uk.westlaw.com/Browse/Home/WestlawUK`,
  },
}

const EU_DATABASES = {
  curia: {
    icon: ``,
    name: `CURIA`,
    url: `https://curia.europa.eu/juris/recherche.jsf?language=en`,
  },
  epo: {
    icon: ``,
    name: `EPO`,
    url: `https://www.epo.org/law-practice/case-law-appeals`,
  },
}

const HK_DATABASES = {
  hkliihk: {
    icon: ``,
    name: `HKLII.hk`,
    url: `https://www.hklii.hk/`,
  },
  hkliiorg: {
    icon: ``,
    name: `HKLII.org`,
    url: `https://www.hklii.org/eng/`,
  },
}

const CA_DATABASES = {
  canlii: {
    icon: ``,
    name: `CanLII`,
    url: `https://www.canlii.org/en`,
  },
}

const AU_DATABASES = {
  austlii: {
    icon: ``,
    name: `AustLII`,
    url: `https://www8.austlii.edu.au`,
  },
}

const MISC_DATABASES = {
  commonlii: {
    icon: ``,
    name: `CommonLII`,
    url: `http://www.commonlii.org`,
  },
}

const dedupeObjects = (inputObject: Record<string, Record<string, string>>, jurisdictionCode: Law.JursidictionCode) =>
  Object.entries(inputObject)
  .reduce((accumulator, [id, object]: [string, Record<string, string>]) => ({
    ...accumulator,
    [`${jurisdictionCode}_${id}`]: object,
  }), {})

const DATABASES: Record<string, Law.Database> = {
  ...dedupeObjects(SG_DATABASES, `SG`),
  ...dedupeObjects(UK_DATABASES, `UK`),
  ...dedupeObjects(EU_DATABASES, `EU`),
  ...dedupeObjects(HK_DATABASES, `HK`),
  ...dedupeObjects(CA_DATABASES, `CA`),
  ...dedupeObjects(AU_DATABASES, `AU`),
  ...MISC_DATABASES,
}

const SG_COURTS = {
  sgca: {
    id: `SGCA`,
    name: `Court of Appeal`,
  },
  sghc: {
    id: `SGHC`,
    name: `High Court`,
  },
}

const EU_COURTS = {
  cjeu: {
    id: `CJEU`,
    name: `Court of Justice of the European Union`,
  },
  epo: {
    id: `EPO`,
    name: `European Patent Office`,
  },
}

const COURTS: Record<string, Record<string, string>> = {
  ...dedupeObjects(SG_COURTS, `SG`),
  ...dedupeObjects(EU_COURTS, `EU`),
}

const Constants = {
  COURTS,
  DATABASES,
  JURISDICTIONS,
}

export default Constants