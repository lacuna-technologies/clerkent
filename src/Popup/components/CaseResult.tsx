import useClipboard from 'Popup/hooks/useClipboard'
import type { FunctionComponent } from 'preact'
import { useCallback } from 'preact/hooks'
import Helpers from 'utils/Helpers'
import ResultLink from './ResultLink'

interface Props {
  case: Law.Case,
  downloadPDF: downloadPDFType
}

const CaseResult: FunctionComponent<Props> = ({
  case: currentCase,
  downloadPDF,
}) => {
  const {
    citation,
    name,
    links,
    database,
  } = currentCase
  const summaryURL = Helpers.getSummaryLink(links)?.url
  const judgmentLink = Helpers.getJudgmentLink(links)
  const opinionLink = Helpers.getOpinionLink(links)
  const orderLink = Helpers.getOrderLink(links)

  const caseNameClass = summaryURL ? `text-blue-700 border-0 bg-none outline-none p-0 underline cursor-pointer select-text hover:text-blue-900 hover:underline` : ``
  const {
    permissionGranted,
    promptGrant,
  } = useClipboard()
  const onClickCitation = useCallback(async () => {
    if(!permissionGranted){
      await promptGrant()
    }
    await navigator.clipboard.writeText(citation)
  }, [permissionGranted, promptGrant, citation])

  return (
    <div className="w-full outline-none">
      <div className="flex flex-row justify-start items-center mb-0.5">
        {
          database && (
            <a className="text-xs" href={database.url} target="_blank" rel="noreferrer">
              {database.name}
            </a>
          )
        }
      </div>
      <a
        className={`align-left text-lg ${caseNameClass}`}
        target="_blank"
        href={summaryURL}
        rel="noreferrer"
        {...(summaryURL ? {title: `Summary`} : {})}
      >
        {name}
      </a>
      <div className="flex flex-row mt-0.5 gap-4">
        <span
          className="cursor-pointer select-all"
          onClick={onClickCitation}
          title="Click to copy citation"
        >
          {citation}
        </span>
        <ResultLink
          link={judgmentLink}
          onDownloadPDF={downloadPDF({ doctype: `Judgment`, law: currentCase })}
        />
        <ResultLink
          link={opinionLink}
          onDownloadPDF={downloadPDF({ doctype: `Opinion`, law: currentCase })}
        />
        <ResultLink
          link={orderLink}
          onDownloadPDF={downloadPDF({ doctype: `Order`, law: currentCase })}
        />
      </div>
    </div>
  )
}

export default CaseResult