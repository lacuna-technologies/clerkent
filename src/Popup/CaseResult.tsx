import React from 'react'
import Constants from '../utils/Constants'

const CaseResult = ({
  citation,
  name,
  link,
  pdf,
  jurisdiction,
  database,
  downloadPDF,
}) => {
  return (
    <div className="result">
      <div className="details">
        <div className="left">
          <span className="jurisdiction">{Constants.JURISDICTIONS[jurisdiction]?.emoji}</span>
          { database && <span className="database">{database.name}</span> }
        </div>
        <div className="right">
          {
            pdf ? (
              <p className="links">
                <button className="pdf button" onClick={downloadPDF({ citation, name, pdf })}>PDF</button>
              </p>
            ) : null
          }
        </div>
      </div>
      <a
        className="case-name link"
        target="_blank"
        href={link}
        rel="noreferrer"
      >
        {name} {citation}
      </a>
    </div>
  )
}

export default CaseResult