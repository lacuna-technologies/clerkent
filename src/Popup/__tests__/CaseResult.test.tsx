import { render } from '@testing-library/preact'
import Constants from 'utils/Constants'
import CaseResult from '../components/CaseResult'
import useClipboard from 'Popup/hooks/useClipboard'

jest.mock(`Popup/hooks/useClipboard`)
const useClipboardMock = useClipboard as jest.Mock

const mockCase: Law.Case = {
  citation: `[2006] EWCA Civ 145`,
  database: Constants.DATABASES.UK_bailii,
  jurisdiction: Constants.JURISDICTIONS.UK.id,
  links: [
    {
      doctype: `Judgment`,
      filetype: `HTML`,
      url: `https://www.bailii.org/ew/cases/EWCA/Civ/2006/145.html`,
    },
  ],
  name: `IDA v University of Southampton`,
  type: `case-name`,
}
const mockDownloadPDF = () => () => null

describe(`CaseResult`, () => {
  beforeAll(() => {
    useClipboardMock.mockImplementation(() => ({
      permissionGranted: true,
      promptGrant: jest.fn(),
    }))
  })
  it(`renders without error`, () => {
    const tree = render(
      <CaseResult
        downloadPDF={mockDownloadPDF}
        case={mockCase}
      />,
    )
    expect(tree).toMatchSnapshot()
  })
})