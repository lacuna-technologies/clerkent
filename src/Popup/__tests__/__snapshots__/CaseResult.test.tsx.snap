// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CaseResult renders without error 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="w-full outline-none"
      >
        <div
          class="flex flex-row justify-start items-center mb-0.5"
        >
          <a
            class="text-xs"
            href="https://www.bailii.org/"
            rel="noreferrer"
            target="_blank"
          >
            BAILII
          </a>
        </div>
        <a
          class="align-left text-lg "
          rel="noreferrer"
          target="_blank"
        >
          IDA v University of Southampton
        </a>
        <div
          class="flex flex-row mt-0.5 gap-4"
        >
          <span>
            [2006] EWCA Civ 145
          </span>
          <div
            class="link judgment flex flex-row text-blue-700 border-0 bg-none outline-none p-0 underline cursor-pointer select-text hover:text-blue-900"
          >
            <a
              href="https://www.bailii.org/ew/cases/EWCA/Civ/2006/145.html"
              rel="noreferrer"
              target="_blank"
            >
              Judgment
            </a>
            <a
              href="https://www.bailii.org/ew/cases/EWCA/Civ/2006/145.html"
            >
              <img
                alt="download PDF"
                class="h-4 ml-2 mt-[0.3rem] pdf-filter"
                src="{\\"path\\":\\"pdf.svg\\",\\"source\\":\\"<svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"16\\" height=\\"16\\" fill=\\"currentColor\\" class=\\"bi bi-file-earmark-pdf\\" viewBox=\\"0 0 16 16\\">
  <path d=\\"M14 14V4.5L9.5 0H4a2 2 0 0 0-2 2v12a2 2 0 0 0 2 2h8a2 2 0 0 0 2-2zM9.5 3A1.5 1.5 0 0 0 11 4.5h2V14a1 1 0 0 1-1 1H4a1 1 0 0 1-1-1V2a1 1 0 0 1 1-1h5.5v2z\\"/>
  <path d=\\"M4.603 14.087a.81.81 0 0 1-.438-.42c-.195-.388-.13-.776.08-1.102.198-.307.526-.568.897-.787a7.68 7.68 0 0 1 1.482-.645 19.697 19.697 0 0 0 1.062-2.227 7.269 7.269 0 0 1-.43-1.295c-.086-.4-.119-.796-.046-1.136.075-.354.274-.672.65-.823.192-.077.4-.12.602-.077a.7.7 0 0 1 .477.365c.088.164.12.356.127.538.007.188-.012.396-.047.614-.084.51-.27 1.134-.52 1.794a10.954 10.954 0 0 0 .98 1.686 5.753 5.753 0 0 1 1.334.05c.364.066.734.195.96.465.12.144.193.32.2.518.007.192-.047.382-.138.563a1.04 1.04 0 0 1-.354.416.856.856 0 0 1-.51.138c-.331-.014-.654-.196-.933-.417a5.712 5.712 0 0 1-.911-.95 11.651 11.651 0 0 0-1.997.406 11.307 11.307 0 0 1-1.02 1.51c-.292.35-.609.656-.927.787a.793.793 0 0 1-.58.029zm1.379-1.901c-.166.076-.32.156-.459.238-.328.194-.541.383-.647.547-.094.145-.096.25-.04.361.01.022.02.036.026.044a.266.266 0 0 0 .035-.012c.137-.056.355-.235.635-.572a8.18 8.18 0 0 0 .45-.606zm1.64-1.33a12.71 12.71 0 0 1 1.01-.193 11.744 11.744 0 0 1-.51-.858 20.801 20.801 0 0 1-.5 1.05zm2.446.45c.15.163.296.3.435.41.24.19.407.253.498.256a.107.107 0 0 0 .07-.015.307.307 0 0 0 .094-.125.436.436 0 0 0 .059-.2.095.095 0 0 0-.026-.063c-.052-.062-.2-.152-.518-.209a3.876 3.876 0 0 0-.612-.053zM8.078 7.8a6.7 6.7 0 0 0 .2-.828c.031-.188.043-.343.038-.465a.613.613 0 0 0-.032-.198.517.517 0 0 0-.145.04c-.087.035-.158.106-.196.283-.04.192-.03.469.046.822.024.111.054.227.09.346z\\"/>
</svg>\\"}"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="w-full outline-none"
    >
      <div
        class="flex flex-row justify-start items-center mb-0.5"
      >
        <a
          class="text-xs"
          href="https://www.bailii.org/"
          rel="noreferrer"
          target="_blank"
        >
          BAILII
        </a>
      </div>
      <a
        class="align-left text-lg "
        rel="noreferrer"
        target="_blank"
      >
        IDA v University of Southampton
      </a>
      <div
        class="flex flex-row mt-0.5 gap-4"
      >
        <span>
          [2006] EWCA Civ 145
        </span>
        <div
          class="link judgment flex flex-row text-blue-700 border-0 bg-none outline-none p-0 underline cursor-pointer select-text hover:text-blue-900"
        >
          <a
            href="https://www.bailii.org/ew/cases/EWCA/Civ/2006/145.html"
            rel="noreferrer"
            target="_blank"
          >
            Judgment
          </a>
          <a
            href="https://www.bailii.org/ew/cases/EWCA/Civ/2006/145.html"
          >
            <img
              alt="download PDF"
              class="h-4 ml-2 mt-[0.3rem] pdf-filter"
              src="{\\"path\\":\\"pdf.svg\\",\\"source\\":\\"<svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"16\\" height=\\"16\\" fill=\\"currentColor\\" class=\\"bi bi-file-earmark-pdf\\" viewBox=\\"0 0 16 16\\">
  <path d=\\"M14 14V4.5L9.5 0H4a2 2 0 0 0-2 2v12a2 2 0 0 0 2 2h8a2 2 0 0 0 2-2zM9.5 3A1.5 1.5 0 0 0 11 4.5h2V14a1 1 0 0 1-1 1H4a1 1 0 0 1-1-1V2a1 1 0 0 1 1-1h5.5v2z\\"/>
  <path d=\\"M4.603 14.087a.81.81 0 0 1-.438-.42c-.195-.388-.13-.776.08-1.102.198-.307.526-.568.897-.787a7.68 7.68 0 0 1 1.482-.645 19.697 19.697 0 0 0 1.062-2.227 7.269 7.269 0 0 1-.43-1.295c-.086-.4-.119-.796-.046-1.136.075-.354.274-.672.65-.823.192-.077.4-.12.602-.077a.7.7 0 0 1 .477.365c.088.164.12.356.127.538.007.188-.012.396-.047.614-.084.51-.27 1.134-.52 1.794a10.954 10.954 0 0 0 .98 1.686 5.753 5.753 0 0 1 1.334.05c.364.066.734.195.96.465.12.144.193.32.2.518.007.192-.047.382-.138.563a1.04 1.04 0 0 1-.354.416.856.856 0 0 1-.51.138c-.331-.014-.654-.196-.933-.417a5.712 5.712 0 0 1-.911-.95 11.651 11.651 0 0 0-1.997.406 11.307 11.307 0 0 1-1.02 1.51c-.292.35-.609.656-.927.787a.793.793 0 0 1-.58.029zm1.379-1.901c-.166.076-.32.156-.459.238-.328.194-.541.383-.647.547-.094.145-.096.25-.04.361.01.022.02.036.026.044a.266.266 0 0 0 .035-.012c.137-.056.355-.235.635-.572a8.18 8.18 0 0 0 .45-.606zm1.64-1.33a12.71 12.71 0 0 1 1.01-.193 11.744 11.744 0 0 1-.51-.858 20.801 20.801 0 0 1-.5 1.05zm2.446.45c.15.163.296.3.435.41.24.19.407.253.498.256a.107.107 0 0 0 .07-.015.307.307 0 0 0 .094-.125.436.436 0 0 0 .059-.2.095.095 0 0 0-.026-.063c-.052-.062-.2-.152-.518-.209a3.876 3.876 0 0 0-.612-.053zM8.078 7.8a6.7 6.7 0 0 0 .2-.828c.031-.188.043-.343.038-.465a.613.613 0 0 0-.032-.198.517.517 0 0 0-.145.04c-.087.035-.158.106-.196.283-.04.192-.03.469.046.822.024.111.054.227.09.346z\\"/>
</svg>\\"}"
            />
          </a>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
